# Ejercicio 1

> Por: Mapachana

**Dado tu número n = d1d2d3d4d5d6d7d8.**

En este caso mi número es $n=77770080$.

In [1]:
n = 77770080

print(n)

77770080


**i) Mientras n sea múltiplo de 2, 3, 5, 7 u 11 le sumas uno. De forma que tu nuevo nno tenga esos divisores primos.**

In [2]:
while n%2 == 0 or n%3 == 0 or n%5 == 0 or n%7 == 0 or n%11 == 0:
    n += 1
    
print(n)

77770081


Luego el número con el que vamos a trabajar es $77770081$

**ii) Calcula $a^{n-1}$ mod n, para cada uno de esas cinco bases usando sucesivamente elalgoritmo de izda-drcha y de drcha-izda.**

Comenzamos implementando el algoritmo de exponenciación rápida de derecha a izquierda, ya que es más intuitivo de entender:

In [3]:
def exp_dcha_izq(a, b, n):
    '''Funcion que calcula mediante exponenciacion rapida de derecha a izquierda a^b mod n'''
    base = a
    acu = 1
    num = b
    exp = 0
    
    # Iteracion 0
    print("  Paso  |  acu  |  base  ")
    print("-------------------------")
    print("  0  |   {}  |  {}  ".format(acu, base))
    
    j = 1
    while num > 0:
        if num % 2 != 0:
            acu = (acu * base) % n
            num = (num-1)/2
            ej = 1
        else:
            num = num/2
            ej = 0
        base = (base*base) % n
        exp = exp + ej*pow(2, j)

        print("  {}  |  {}  |  {}  ".format(j, acu, base))
        
        j += 1

In [4]:
# Numeros ejemplo diapos
exp_dcha_izq(2, 27213646, 27213647)

  Paso  |  acu  |  base  
-------------------------
  0  |   1  |  2  
  1  |  1  |  4  
  2  |  4  |  16  
  3  |  64  |  256  
  4  |  16384  |  65536  
  5  |  16384  |  22424717  
  6  |  16384  |  12167649  
  7  |  14796941  |  26585339  
  8  |  14796941  |  9779482  
  9  |  120175  |  15067521  
  10  |  24905736  |  12424177  
  11  |  9320126  |  14159809  
  12  |  10785431  |  15283  
  13  |  682094  |  15860913  
  14  |  11508854  |  2314287  
  15  |  11508854  |  6452299  
  16  |  11508854  |  17649979  
  17  |  15616437  |  11515927  
  18  |  8921297  |  9800221  
  19  |  10593740  |  23701151  
  20  |  1995822  |  21932449  
  21  |  4173990  |  3214080  
  22  |  4173990  |  9845200  
  23  |  4173990  |  760573  
  24  |  16105485  |  18007697  
  25  |  1  |  13932984  


In [5]:
lista_num = [2, 3, 5, 7, 11]

for num in lista_num:
    print("Para la base {} se tiene: ".format(num))
    exp_dcha_izq(num, n-1, n)


Para la base 2 se tiene: 
  Paso  |  acu  |  base  
-------------------------
  0  |   1  |  2  
  1  |  1  |  4  
  2  |  1  |  16  
  3  |  1  |  256  
  4  |  1  |  65536  
  5  |  1  |  17612841  
  6  |  17612841  |  69275565  
  7  |  44516305  |  9980674  
  8  |  44516305  |  23223320  
  9  |  44680160  |  67690927  
  10  |  44680160  |  77257279  
  11  |  1465333  |  25247343  
  12  |  70708033  |  13797891  
  13  |  70708033  |  15599233  
  14  |  31848612  |  19426093  
  15  |  31848612  |  70568710  
  16  |  19282073  |  10086087  
  17  |  19282073  |  52267494  
  18  |  44352822  |  10929208  
  19  |  44352822  |  53478878  
  20  |  44352822  |  18488420  
  21  |  44352822  |  3628315  
  22  |  35935437  |  61507869  
  23  |  35935437  |  45590014  
  24  |  15507894  |  21699431  
  25  |  15507894  |  50932700  
  26  |  15507894  |  44851232  
  27  |  1  |  55498452  
Para la base 3 se tiene: 
  Paso  |  acu  |  base  
-------------------------
  0  |   

Implementamos ahora la función de izquierda a derecha

In [6]:
def calcular_digitos(b):
    digitos = []
    
    while b != 1:
        resto = b%2
        digitos.append(resto)
        b = (b-resto)/2
    
    digitos.append(1)
        
    return digitos
        

def exp_izq_dcha(a, b, n):
    '''Funcion que calcula mediante exponenciacion rapida de izquierda a derecha a^b mod n'''
    acu = 1
    exp = 0
    
    digitos = calcular_digitos(b)
    
    # Iteracion 0
    print("  Paso  |  acu  ")
    print("-------------------------")
    print("  0  |   {} ".format(acu))
    
    j = 1
    digitos = list(reversed(digitos))
    while j < len(digitos)+1:
        acu = acu*acu
        exp = exp*2
        
        ej = digitos[j-1]
        if ej == 1:
            acu = acu*a
            exp += 1
        
        acu = acu % n

        print("  {}  |  {} ".format(j, acu))
        
        j += 1

In [7]:
# Numero ejemplo diapos
exp_izq_dcha(2, 27213646, 27213647)

  Paso  |  acu  
-------------------------
  0  |   1 
  1  |  2 
  2  |  8 
  3  |  64 
  4  |  4096 
  5  |  6340785 
  6  |  6967674 
  7  |  2913255 
  8  |  5278505 
  9  |  8390032 
  10  |  21530122 
  11  |  20718507 
  12  |  6234754 
  13  |  14354021 
  14  |  18664779 
  15  |  21238793 
  16  |  9159785 
  17  |  1897988 
  18  |  6353813 
  19  |  22747524 
  20  |  12082479 
  21  |  22922644 
  22  |  15226206 
  23  |  8424067 
  24  |  1 
  25  |  1 


In [8]:
lista_num = [2, 3, 5, 7, 11]

for num in lista_num:
    print("Para la base {} se tiene: ".format(num))
    exp_izq_dcha(num, n-1, n)

Para la base 2 se tiene: 
  Paso  |  acu  
-------------------------
  0  |   1 
  1  |  2 
  2  |  4 
  3  |  16 
  4  |  512 
  5  |  262144 
  6  |  19220345 
  7  |  11864688 
  8  |  53190135 
  9  |  72776920 
  10  |  15099639 
  11  |  7072054 
  12  |  17375632 
  13  |  33857947 
  14  |  56968058 
  15  |  75776703 
  16  |  20434621 
  17  |  1784769 
  18  |  15635682 
  19  |  49254338 
  20  |  41393140 
  21  |  65775228 
  22  |  42089411 
  23  |  64834401 
  24  |  77611099 
  25  |  77770080 
  26  |  1 
  27  |  1 
Para la base 3 se tiene: 
  Paso  |  acu  
-------------------------
  0  |   1 
  1  |  3 
  2  |  9 
  3  |  81 
  4  |  19683 
  5  |  76340165 
  6  |  19702455 
  7  |  17898279 
  8  |  51068909 
  9  |  34810944 
  10  |  23313615 
  11  |  47220160 
  12  |  20151576 
  13  |  49392394 
  14  |  9595127 
  15  |  62696061 
  16  |  68121819 
  17  |  39666683 
  18  |  61569031 
  19  |  70952581 
  20  |  28351403 
  21  |  64679486 
  22  |  77

**iii) ¿Es n un posible primo de Fermat para alguna de ellas? ¿Es n un pseudoprimopara alguna de ellas?**

Del apartado anterior tenemos que $2^{n-1}\equiv 1 \mod n$, $3^{n-1}\equiv 1 \mod n$ , $5^{n-1}\equiv 1 \mod n$, $7^{n-1}\equiv 1 \mod n$ y $11^{n-1}\equiv 1 \mod n$, luego $n$ es un pseudoprimo de Fermat para todas estas bases.